Changed @author and @since to build a changelog

docbill opened the pull request:
I need to find the procedure to install and test.

closes #51

docbill commented:
The following is what it looks like with this update.
![Screenshot from 2022-09-21 14-58-57](https://user-images.githubusercontent.com/3169368/191663409-cfabcbaa-9d5f-4b13-8eed-c1876e5c84de.png)

And this is what it looked like before hand:
![Screenshot from 2022-09-22 01-10-35](https://user-images.githubusercontent.com/3169368/191663260-72a3fb2e-7638-4fdc-b66b-a1111933bbc6.png)

The @author and @since attributes now correctly build a change log.

docbill commented:
@no-stack-dub-sack Please let me know you want me to walk through the change with you on a google chat call or such.

dschach commented:
It's best to include README and documentation updates with this change, since it appears to be significant

no-stack-dub-sack commented:
@docbill Please see if this comment helps you with testing these changes locally: https://github.com/no-stack-dub-sack/apexdox-vs-code/pull/20#issuecomment-658339950. Though it looks like from your screenshot above you were able to run apex-dox with your changes?

docbill commented:
Yes.  Basically I follewed the same set-up steps repeatedly like an insane person, and eventually it successfully installed and worked.  I'm not sure what mistake I was making initially.

no-stack-dub-sack commented:
@docbill cool, so it looks like your changes are working as expected. ideally we'd add at least a couple of tests to cover this new behavior. even I need to refresh myself on how to go about that at this point (the last time I added any meaningful tests was years ago). Are you willing to give this a shot if I can provide some guidance?

docbill commented:
> It's best to include README and documentation updates with this change, since it appears to be significant

The README does not seem to get into that level of detail on how each tag works...

docbill commented:
> @docbill cool, so it looks like your changes are working as expected. ideally we'd add at least a couple of tests to cover this new behavior. even I need to refresh myself on how to go about that at this point (the last time I added any meaningful tests was years ago). Are you willing to give this a shot if I can provide some guidance?

Sure.  I looked though all the tests, and I could not find any that test @author or @since, but I noticed most of the classes called by these tests include these attributes.   So I updated the classes to include a second set of entries as a change log class.

Before I go through the effort of created a brand new test, I want to make certain I'm not simply missing the relevant tests that I should update.   Ideally what a test should check is typical usage.  And usage where there is only one tag but not the other.

no-stack-dub-sack commented:
@docbill Sorry that my replies are far between, things have been very hectic for me lately, but I have not forgotten about this.

I took a look at the tests as well and refreshed my memory a bit. For the most part I think tag behavior is mostly covered by the snapshot tests, so the changes you've made by updating the test project's classes and then updating the snapshots is likely sufficient. The goal is just to make sure that the comments produce the expected output, so if we manually review the output and it looks good, it should be enough to just update the snapshots. From there we should be able to catch any changes that affect this feature in the future. It's not the most robust testing framework, but its something.

Thank you, btw, you are the first person to be able to figure all the testing stuff out on your own 👍 

As far as the README goes, I do agree with @dschach that since this is introducing a fairly substantial change in behavior that we should probably update the README. Can you please add some description to both of these entries that describes how the two tags work together?

![image](https://user-images.githubusercontent.com/18563015/195230072-5d08e111-6877-4ff8-886c-c087f651c361.png)

Maybe add to both something like: "This tag works with the <other> tag to form a change log", and perhaps an example in the tips section to add a tip that shows a comment block that utilities this feature and what it outputs. Feel free to just manually edit some HTML to produce the expected output from the sample comment.

How does this sound?

Once this is done, I'll just play around with it locally a bit, and then I think this should be good to go. Thanks for the effort put into this!

no-stack-dub-sack reviewed (changes requested):
@docbill Sorry that my replies are far between, things have been very hectic for me lately, but I have not forgotten about this.

I took a look at the tests as well and refreshed my memory a bit. For the most part I think tag behavior is mostly covered by the snapshot tests, so the changes you've made by updating the test project's classes and then updating the snapshots is likely sufficient. The goal is just to make sure that the comments produce the expected output, so if we manually review the output and it looks good, it should be enough to just update the snapshots. From there we should be able to catch any changes that affect this feature in the future. It's not the most robust testing framework, but its something.

Thank you, btw, you are the first person to be able to figure all the testing stuff out on your own 👍 

As far as the README goes, I do agree with @dschach that since this is introducing a fairly substantial change in behavior that we should probably update the README. Can you please add some description to both of these entries that describes how the two tags work together?

![image](https://user-images.githubusercontent.com/18563015/195230072-5d08e111-6877-4ff8-886c-c087f651c361.png)

Maybe add to both something like: "This tag works with the <other> tag to form a change log", and perhaps an example in the tips section to add a tip that shows a comment block that utilities this feature and what it outputs. Feel free to just manually edit some HTML to produce the expected output from the sample comment.

How does this sound?

Once this is done, and the other small change is taken care of, I'll just play around with it locally a bit, and then I think this should be good to go. Thanks for the effort put into this!

**EDIT:** Oh, I also just realized that `@author` and `@since` are also valid method tags. Can you also please verify that it works as expected for methods as well and maybe add an example to one or two of the test classes (it doesn't need to be done everywhere)?

docbill commented on src/common/models/ApexModel.ts:
```diff
@@ -5,11 +5,12 @@ import { existsSync } from 'fs';
 import { resolve } from 'path';
 import { window, workspace, WorkspaceFolder } from 'vscode';
 import { Option } from '../..';
+import { listenerCount } from 'cluster';
```
Yeah.  I'm not sure how this line got added, I'll remove it.

docbill commented:
Upon reviewing the way the changelog looks on methods, I made some revisions.  Originally I was formatting the changelogs simmilar to doxygen when processing javadox.  But since it has been years since I have parsed code with doxygen I may have deviated significantly from that model.

![Screenshot from 2022-10-12 11-20-16](https://user-images.githubusercontent.com/3169368/195384188-f9ee4533-d0a8-4c86-80c3-d8a2ea511017.jpeg)

![Screenshot from 2022-10-12 11-19-13](https://user-images.githubusercontent.com/3169368/195384330-242d6478-a292-4ce5-a81f-ceadee49c906.jpeg)

docbill commented:
I have finished updating the README as well.  One more screenshot worth sharing is for the annotations on TEST_Asnnotations.cls:

![Screenshot from 2022-10-12 12-19-43](https://user-images.githubusercontent.com/3169368/195396214-5cae6987-0ce8-440c-b2af-40614ce8ccd5.jpeg)

This shows how we display and `@since` without an `@author` and then an `@author` without a `@since`.

no-stack-dub-sack commented:
@docbill cool, these screenshots look good to me. I'll take one final look and merge this within the next few days. Thanks for your patience on this!

docbill commented:
Take your time.  I appreciate it is better to carefully review changes to keep up consistent standards.

no-stack-dub-sack reviewed (changes requested):
A couple of minor documentation changes in the README. Also, can you please add `.pmdCache` to .gitignore and remove the file from version control?

More importantly, though...

Now that you've made some additional changes and I've had a chance to really play with this, I do see at least one issue with this that warrants a discussion, and that's that the order of the tags matters. For example, the following comment block:

```java
/**
 * This class should be included by `apexdox.engine.includes` setting
 * 
 * @author Peter Weinberg
 * @since 2019-01-01 Created
 * 
 * @author Bill C Riemers
 * @since 2022-10-05 Added changelog
 */
public class IncludeOne {

}
```
Produces the following, as expected:

![image](https://user-images.githubusercontent.com/18563015/198856231-8a1c581d-1a19-4af6-9d80-f69d033c6bf4.png)

However, if you reverse the tags like so:

```java
/**
 * This class should be included by `apexdox.engine.includes` setting
 * 
 * @since 2019-01-01 Created
 * @author Peter Weinberg
 * 
 * @since 2022-10-05 Added changelog
 * @author Bill C Riemers
 */
public class IncludeOne {

}
```
You get:

![image](https://user-images.githubusercontent.com/18563015/198856260-ad7af238-208d-4f2d-bebe-c0ffc57cb895.png)

Now it takes the middle two tags as a changelog pair, and incorrectly isolates the last `@author` and the first `@since`.

We could just say that it's always expected that `@author` will come before `@since` (which makes sense since `@since` may keep getting updated over time, and I noticed you did already say that they need to be sequenced correctly, but I didn't realize the consequences of not), but it's not totally unreasonable to think that someone may choose to go in the opposite order, and if so, they would get some pretty unexpected results. Can you think of another design approach that may solve this, or are you aware of how doxygen or javadoc handle this?

Off hand, I can't really think of a great solution. I was thinking that maybe building an array of `{ author: string, since: string }` or `{ author: string, since: string[] }`  objects could help, but since we're attempting to support multiple instances of both tags, it gets pretty tricky to determine how the author of the comment intended that they be grouped without insisting that users always use a specific order or face incorrect and most-likely misleading results. I was looking into JavaDoc author and since annotations and they don't really seem to work the same way so far as I can tell. `@since` specifically refers to the version number that a feature or class was implemented in, and `@author` is just creates a list of all contributing authors. Salesforce is of course a bit different when it comes to versioning, so I'm not really sure the best way to improve this. What are your thoughts?

I'm hesitant to include something that is nearly sure to result in bug reports, but I don't think the user-base is _that_ big, and this is the first time this feature was requested, which leads me to believe not many people will be using it this way... so perhaps just a strong warning that using it in the wrong order will produce unexpected results?? Not sure on this one... if you can think of a better design that would at least help to address this, let me know. 

**EDIT:** Maybe the way to do it is to collect all `@author` and `@since` tags that immediately follow one another and treat them as a single group, so that groups separated by a blank line will be processed independently of one another. e.g.

```
@since Date1
@since Date2
@author Pete
@author Bill

@author John
@since Date 3
```
and the same groups re-ordered will always produce the same markup which could look something like:

```
Author: Pete, Bill
  - Date 1
  - Date 2

Author: John
  - Date 3
```

By creating an array of `{ author: string[], date: string[] }` for each group separated by a white line, we could then produce the markup appropriately for each group. What do you think? Then we just have to let users know that a blank line between groups of these tags will create a new entry in the changelog.

docbill commented on README.md:
```diff
@@ -180,6 +180,23 @@ Located in the lines above any top-level type that lives within a .cls file, or
     </a>
 </p>
 
+You can create a full revision history by sequencing @author and @since attributes appropriately:
```
I would expect:

```
/**
 * This class should be included by `apexdox.engine.includes` setting
 * 
 * @since 2019-01-01 Created
 * @author Peter Weinberg
 * 
 * @since 2022-10-05 Added changelog
 * @author Bill C Riemers
 */
```

To parse exactly the way it does.   The reason being, is Apex, and ApexDox both work in top down order.  If I write:

```
/**
 * Integer foo = bar();
 * @example
 */
```
I don't expect `Integer foo = bar()` to appear in the **example** section.  Because the attribute comes after that text not before.

And in general @ApexDox does not treat blank lines special.  So I would expect it to parse:

```
/**
 * This class should be included by `apexdox.engine.includes` setting
 * 
 * @since 2019-01-01
 * 
 * @author Peter Weinberg
 * @since 2022-10-05 Added changelog
 * 
 * @author Bill C Riemers
 */
```
Exactly the same.  It is obvious to you and me, that in this case the first  `@since` is being used as a version tag, you would have been the one that made the change on 2022-10-05, and I never bothered to indicate a comment or date when I created the class.

But I don't expect the parser to be smart enough to distiquish between these two cases.  In fact I would be very concerned if it did, as it would mean the results would be more prone to unpredictable results.

And actually I do think there is a valid use case where `@author` will come before `@since`.   Any documention created before this update will likely just have one `@author` tag, and one `@since` in random order.   If you look at the test files, I deliberately put the tags out of order in one case, just to make sure the results seem reasonable to me.

docbill replied:
One thing though that kept coming to mind, is maybe it would be a good idea to have a configuration option to have the old ungrouped parsing of @author and @since.   But I really couldn't come up with a good example of where I thought that was needed, so it seemed like a way to just make it more complicated.

docbill replied:
Admittedly though, I cannot find anywhere that even documents doxygen works this way.  It just does...

The idea of using `@since` and specifying only a date never made sense to me.  As it is like someone suddenly standing up in the room and shouting a date.   If they don't follow that up about what is significant about that date, I'm likely to dismiss them as a lunatic.   So I was not at all all surprised when I found examples of people using it as a change log in this manner.  But that was at least 12 years ago.  I'd be hard pressed to find those samples again today.

docbill replied:
I did however, knowingly deviate from even that model abit.  In that if I specify:

```
* @author Bill Riemers
* @author Liam O'Connor
* @since 2021-10-22 Created
* @since 2022-10-21 Added investigations method
* 
* @author Buffy Summers
* @since 2022-01-13 Staked things
```

doxygen would give combine Bill Riemers and Liam O'Connor into a column separated list, under the assumption the both worked the changes that follow together.   

I did not attempt to reproduce that behavour mainly because out of the millions of characters of code Red Hat and our contractors have put in salesforce, not a single developer has found it necessary to list two authors in a row.  I notice even in the cases I know two developer pair coded a solution together, they each add there own `@since` statements.

But if we want to merge the authors together in this case, it wouldn't be to difficult to change the code to do that.  I personally like incremental changes.  So my suggestion is if we want to do that, I do a follow-up request.

docbill replied:
A slightly better idea.  I make a class for ordered attributes: { name: String, value: String [] }.  So changelog would be an array of ordered attributes.   If the attribute is the same name as the previous, we append to the value array of the last ordered attributes.  If the attribute is different we append a new entry to the ordered attributes array.   Then output could look something like:

**Author(s):** **Peter Weinberg**, **Bill Riemers**
**Since:**
- 2022-10-22 - Pair coded apple attribute
- 2022-10-28 - Added orange attribute
**Author(s):** **Peter Weinberg**
**Since:** 2021-09-15 - Created

This is actually order independant, because we are just outputting in the same order specified.  It is up to the reader to interpret if there is significance to this order.

Notice: In this case I would make `@since` a list only if there is more than one value.  I think this styling could all be done in css, but it has been a few years since I had to do that in css, so if we want that it would probably be a follow-up as it would take several iterations to get it looking really nice.

no-stack-dub-sack replied:
@docbill I'm ok with this suggestion, and I think as long as we're not introducing any behavior that's too unpredictable I'm good with it. I also love the markdown idea, but I do think there's already an extension out there that does it. Let me see if I can find it for you...

no-stack-dub-sack replied:
Also, are you able to run the extension without packaging it first? i.e. in debug mode? I got a new machine since I last worked on this and due to some updates to how vscode extensions are built since then and using the latest version of VS Code I was really struggling to get the extension working in debug mode. In the interest of time I just ended up packaging it and installing from vsix to test out the changes locally.

no-stack-dub-sack replied:
Hate to send you away from my extension, but if you're looking for markdown, perhaps this would work better for you? https://github.com/cesarParra/apexdocs

docbill replied:
![Screenshot from 2022-10-31-2](https://user-images.githubusercontent.com/3169368/199051320-b6f3e38d-57c1-4041-9e6b-6d69f44751d8.png)
![Screenshot from 2022-10-31-1](https://user-images.githubusercontent.com/3169368/199051349-c1a5ee0d-f6ee-4fe2-b6b8-fd7a10648fe8.png)

docbill replied:
It is not quite as pretty as a change log.  But the code is simple, and it takes care of the order dependence by simply preserving order, and not assuming the order has any particular meaning.   Other than of course we are grouping like tags together only until there is a non-like tag.

docbill replied:
In regards to running in debugging.  I managed to succeed in that once, and having managed to do it again using what I thought was the exact same sequence.  However, the one time was enough for me to figure out the usage of the snapshots.

no-stack-dub-sack commented:
@docbill I'm good with these changes. I made some small updates to syntax and changed `Author(s)` to just `Author` if there's only one author and `Authors` if there's more than one. I also fixed a couple of broken tests that began failing when you introduced `method14` in one of the test classes and updated the snapshots a final time. Everything appears to be in order so I'm going to go ahead and merge, package, and publish. We'll see how it goes out in the wild. Thanks for your effort on this and for your patience. Things have been very hectic for me lately. Hopefully I'll get to some of the other open PRs soon.

no-stack-dub-sack commented:
Oh, and I updated the actual changelog :-)

no-stack-dub-sack commented:
@docbill https://github.com/no-stack-dub-sack/apexdox-vs-code/releases/tag/v1.5.1
